Paperwork clone proc

PsyCommando opened the pull request:
## Description of changes
Depends on #2552. 
Implement the Clone() proc at the datum level, so it can be standardized and used across all entities as needed.
This is mainly a stub implementation mainly for paperwork items, reagent datum, dna datum, and other datum that used a proc named clone() in order to keep things coherent.

## Why and what will this PR improve
Allow a standardized and overridable way to clone datums/objects. As discussed on #2552.

MistakeNot4892 commented on code/game/dna/dna2.dm:
```diff
@@ -86,21 +86,23 @@ var/global/list/datum/dna/gene/dna_genes[0]
+	copy_instance.real_name      = real_name
+	copy_instance.species        = species || global.using_map.default_species
+	copy_instance.body_markings  = body_markings.Copy()
+	for(var/b=1; b<=DNA_SE_LENGTH; b++)
```
While you're in the region, could you swap this to `var/b in 1 to DNA_SE_LENGTH` - tiny microopt but it bugs me to see C-style loop iterators.

PsyCommando replied:
Alright.

MistakeNot4892 commented on code/game/objects/item.dm:
```diff
@@ -148,6 +148,23 @@
+	if(!copy_instance)
+		copy_instance = new type(null, material)
+	copy_instance.contaminated = contaminated
+	copy_instance.blood_overlay = image(blood_overlay)
```
Does passing an image to an image copy the image?

PsyCommando replied:
Actually, I'm not entirely sure now that you mention it. 🤔 
Also, that's the image proc, not the actual image constructor being called. Which also might act differently.

PsyCommando replied:
Asked on the nebula discord and it should copy it.

MistakeNot4892 commented on code/game/objects/item.dm:
```diff
@@ -148,6 +148,23 @@
 
+/obj/item/Clone(obj/item/copy_instance = null)
+	if(!copy_instance)
+		copy_instance = new type(null, material)
```
`material.type` is needed here I believe.

PsyCommando replied:
Either works, it has a check if its an instance or path.
```dm
/obj/item/Initialize(var/ml, var/material_key)
	if(isnull(health))
		health = max_health //Make sure to propagate max_health to health var before material setup, for consistency
	if(!ispath(material_key, /decl/material))
		material_key = material
	if(material_key)
		set_material(material_key)
```

MistakeNot4892 replied:
This appears to be incorrect.
````
/obj/item/proc/set_material(var/new_material)
	if(new_material)
		material = GET_DECL(new_material)
````
This is going to null material if an instance is provided to Initialize().

PsyCommando replied:
hmmm.. I'll take another look.

PsyCommando replied:
Alright, I modified GetCloneArgs to return the material type when there's a material set, instead of passing the instance directly.

MistakeNot4892 commented on code/game/objects/objs.dm:
```diff
@@ -263,6 +263,36 @@
+	if(!copy_instance)
+		copy_instance = new type
+	
+	//Obj stuff
```
Should be able to just set `copy_instance.appearance = src` and all these vars will be set.

PsyCommando replied:
I don't know if I trust some eldritch poorly documented feature to do that for me ^^;
Particularly after I had a talk about image copying and the behaviour of the appearance var with the others.

MistakeNot4892 replied:
Appearances are not an eldritch poorly documented feature. I don't know what talk you had but the behavior I'm describing is reliable.

MistakeNot4892 commented on code/modules/modular_computers/file_system/data.dm:
```diff
@@ -7,10 +7,12 @@
-	return temp
+/datum/computer_file/data/Clone(datum/computer_file/data/copy_instance = null, rename = FALSE)
+	if(!copy_instance)
+		copy_instance = new type
```
Why not just enforce parent calls and return `copy_instance`? You're calling parent anyway.

PsyCommando replied:
Because this is more flexible. You can pass any type you want down the chain. So say you want your copies to be of another type that inherits from the same parents, you can just create it from the type you want, and pass it down. 
That could be handy for things like contracts that would all keep a reference to the source contract and cancel themselves when the master contract is cancelled, and stuff like that.

PsyCommando replied:
Also, constructor parameters are validated in many atoms, so you have to be able to override the creation of the instance per subclass.

MistakeNot4892 replied:
This isn't more flexible at all, you've just duolicated a bunch of logic. You can pass a type down with other patterns and not rely on boilerplate in every override, see my suggestion below.

MistakeNot4892 commented on code/datums/datum.dm:
```diff
@@ -67,3 +67,11 @@
+ * copy_instance is an optional parameter with the instance to copy the content of the src datum into. It's expected to be of a subtype of the src of the proc being called on.
+ */
+/datum/proc/Clone(var/datum/copy_instance = null)
+	return
```
Per parent comment further down - could just do `return new type` and use parent calls in all the other logic. It doesn't seem like there's many cases that would want to skip a step in the tree when cloning a datum.

MistakeNot4892 commented on code/modules/reagents/Chemistry-Holder.dm:
```diff
@@ -162,8 +174,14 @@ var/global/obj/temp_reagents_holder = new
 
-/datum/reagents/proc/add_reagent(var/reagent_type, var/amount, var/data = null, var/safety = 0, var/defer_update = FALSE)
+///Set and call updates on the target holder. 
+/datum/reagents/proc/set_holder(var/obj/new_holder)
```
Should abort out early if `my_atom == new_holder`

PsyCommando replied:
True.

MistakeNot4892 commented on code/modules/reagents/Chemistry-Holder.dm:
```diff
@@ -24,6 +24,18 @@ var/global/obj/temp_reagents_holder = new
+
+	for(var/rtype in reagent_volumes)
+		var/list/rdata = LAZYACCESS(reagent_data, rtype)
+		copy_instance.add_reagent(rtype, reagent_volumes[rtype], rdata?.Copy(), TRUE, TRUE)
```
You don't need to do this, and doing it this way causes a ton of unnecessary overhead. The base datum state is all you need, so just grab it directly.
```
copy_instance.maximum_volume = maximum_volume
copy_instance.total_volume = total_volume
copy_instance.reagent_data = reagent_data?.Copy()
copy_instance.reagent_volumes = reagent_volumes?.Copy()
copy_instance.cached_color = cached_color
copy_instance.primary_reagent = primary_reagent
copy_instance.temperature = temperature
```
You don't need to update values or handle updates either, though you may need to queue for reactions.

PsyCommando replied:
That's the first thing I tried. But, I wasn't really sure it was safe to just do that?
Like the ``reagent_data`` list which contains many instances of lists. If you just copy the ``reagent_data`` list like that, you're not copying the instances of each individual lists inside of it, you're just referring to the existing ones, no?

And then, can we be sure that state of the cached data in the reagents datum is up to date always?

MistakeNot4892 replied:
Use deepcopy then, point is the same.
If the cached state is invalid on the source datum then that's already a separate problem.

PsyCommando replied:
That's fair. I'll go with your suggestion then.

MistakeNot4892 commented on code/modules/paperwork/photography.dm:
```diff
@@ -80,11 +80,11 @@
-	return copy
+/obj/item/photo/Clone(obj/item/photo/copy_instance = null)
+	if(!copy_instance)
+		copy_instance = new type(null, material, img, scribble)
```
Consider a `GetCloneArgs()` proc that returns listargs for use in the base Clone proc instead of all this boilerplate instance creation.

MistakeNot4892 commented:
I would suggest something like:
```
/datum/proc/GetCloneArgs()
    return

/datum/proc/PopulateClone(var/datum/clone)
    return clone

/datum/proc/Clone()
    return PopulateClone(new type(GetCloneArgs()))

...

/obj/GetCloneArgs()
    return list(null, material?.type)
 
/obj/someobj/Clone()
    var/obj/someobj/myclone = ..()
    if(myclone)
        myclone.update_icon()
    return myclone
```

This would allow removing a lot of the boilerplate in your current procs.

PsyCommando commented:
Alright, I tried to apply the changes you asked for. Had to implement my own deep copy proc because the current one just recursively copy list and doesn't clone anything else. So I wrote one that cloned datums, and left atom refs as-is.

MistakeNot4892 commented:
What's the logic behind not cloning atoms? No strong feelings either way but it feels strange that the proc wouldn't handle them. I could understand printing a warning about cloning of atoms being unsupported currently, possibly.

MistakeNot4892 commented on code/game/objects/item.dm:
```diff
@@ -148,6 +148,33 @@
 
+/obj/item/GetCloneArgs()
+	. = ..()
+	LAZYADD(., material)
```
I think this has come up before but this should probably be `material.type` and handled on `/obj`.

MistakeNot4892 replied:
Disregard I was thinking of /obj/structure

MistakeNot4892 commented on code/datums/weakref.dm:
```diff
@@ -39,3 +39,10 @@
 	return "[ref_name] ([ref_type]) ([ref]) (WEAKREF)"
+
+//Allow cloning weakrefs
+/weakref/PopulateClone(weakref/clone)
```
I don't think cloning a weakref makes sense conceptually. `weakref(foo)` will always return the same weakref.

PsyCommando replied:
Alright, 
I keep thinking of C++ weak pointers..

PsyCommando replied:
Marked them as not being able to be cloned. So they're only gonna be copied by reference instead.

MistakeNot4892 commented on code/datums/repositories/decls.dm:
```diff
@@ -117,3 +117,7 @@ var/global/repository/decls/decls_repository = new
+
+/decl/Clone()
+	SHOULD_CALL_PARENT(FALSE)
+	return src //Return a reference, since we're a singleton
```
This should print a warning or error I think,

PsyCommando replied:
It used to, but because we're deep copying lists, it very likely we'll hit /decl in them, 
I guess maybe adding a proc to check if something is clonable first, or if its a singleton would make more sense.

PsyCommando replied:
Just leaving this here for making the convo easier to follow, but this was corrected, and now Clone is never allowed to be called on /decl.

MistakeNot4892 commented on code/_helpers/lists.dm:
```diff
@@ -750,6 +750,21 @@ proc/dd_sortedObjectList(list/incoming)
+		return L
+	. = L.Copy()
+	for(var/i = 1 to length(L))
+		if(islist(.[i]))
```
Shouldn't this be deepcopying anything in the list, regardless of whether or not it's a list itself? Otherwise if you pas a list of datum references you will just get the same references in a new list back.

PsyCommando replied:
Woops, yeah, I forgot to have it call deepcopy on lists and datums..

PsyCommando commented:
> What's the logic behind not cloning atoms? No strong feelings either way but it feels strange that the proc wouldn't handle them. I could understand printing a warning about cloning of atoms being unsupported currently, possibly.

Well, I feel it makes sense? Why would you ever want to clone atoms this way? It's not really something you'd want to do haphazardly with a general purpose proc like this? Way too many unknowns that you can't guarantee in this proc.

Also, atoms are owned by the world, so you're not going to get ownership on their instance by copying them. Creating a copy you have full ownership on is essentially what a deep copy is trying to achieve.

MistakeNot4892 commented:
Hence printing a warning. If you're going to arbitrarily not let atoms be cloned, but implement the clone logic on them (as you have), it's going to cause confusion. Currently it looks/feels inconsistent rather than 'making sense' as an observer.

PsyCommando commented:
> Hence printing a warning. If you're going to arbitrarily not let atoms be cloned, but implement the clone logic on them (as you have), it's going to cause confusion. Currently it looks/feels inconsistent rather than 'making sense' as an observer.

If you mean the /decl I changed it. It now crash when calling Clone() on it.
I added a proc CanClone() to check if we can clone something. And the deep copy proc now check this in order to either use a copy or reference when deep copying the list.

PsyCommando commented:
Ok, so I tried something else with the proc. 
I've added an arg to skip cloning atoms subtypes since making a separate proc would be mostly copy-pasting. And I've used the proc in a lot less places. Opting for deepCopyList and just Copy in a bunch of places.

Let me know if that would be good enough for you? I feel its a bit cleaner and gives more control over what the proc is doing.
